consensus/hotstuff/safetyrules/safety_rules.go

```diff
@@ -93,16 +105,76 @@ func (v *SafetyRules) ProduceVote(proposal *model.Proposal, curView uint64) (*mo
+func (r *SafetyRules) ProduceTimeout(curView uint64, highestQC *flow.QuorumCertificate, highestTC *flow.TimeoutCertificate) (*model.TimeoutObject, error) {
+	lastTimeout := r.safetyData.LastTimeout
+	if lastTimeout != nil && lastTimeout.View == curView {
+		return lastTimeout, nil
+	}
+
+	if !r.IsSafeToTimeout(curView, highestQC, highestTC) {
+		return nil, model.NoTimeoutError{Msg: "not safe to time out under current conditions"}
+	}
+
+	timeout, err := r.signer.CreateTimeout(curView, highestQC, highestTC)
+	if err != nil {
+		return nil, fmt.Errorf("could not timeout at view %d: %w", curView, err)
+	}
+
+	r.safetyData.HighestAcknowledgedView = curView
+	r.safetyData.LastTimeout = timeout
+
+	err = r.persist.PutSafetyData(r.safetyData)
+	if err != nil {
+		return nil, fmt.Errorf("could not persist safety data: %w", err)
+	}
+
+	return timeout, nil
 }
 
 // IsSafeToVote checks if this proposal is valid in terms of voting rules, if voting for this proposal won't break safety rules.
-func (v *SafetyRules) IsSafeToVote(proposal *model.Proposal) bool {
-	panic("to be implemented")
+func (r *SafetyRules) IsSafeToVote(proposal *model.Proposal) error {
+	blockView := proposal.Block.View
+	qcView := proposal.Block.QC.View
+
+	// block's view must be larger than the view of the included QC
+	if blockView <= qcView {
+		return fmt.Errorf("block's view %d must be larger than the view of the included QC %d", blockView, qcView)
+	}
+
+	// block's view must be greater than the view that we have voted for
+	if proposal.Block.View <= r.safetyData.HighestAcknowledgedView {
+		return model.NoVoteError{Msg: "not safe to vote, we have already voted for this view"}
+	}
+
+	// This check satisfies voting rule 1 and 2a.
+	if blockView == qcView+1 {
+		return nil
+	}
+
+	// This check satisfies voting rule 1 and 2b.
+	lastViewTC := proposal.LastViewTC
+	if lastViewTC != nil {
+		if blockView == lastViewTC.View+1 {
+			if qcView >= lastViewTC.TOHighestQC.View {
+				return nil
+			} else {
+				return fmt.Errorf("QC's view %d should be at least %d", qcView, lastViewTC.TOHighestQC.View)
+			}
+		} else {
+			return fmt.Errorf("last view TC %d is not sequential for block %d", lastViewTC.View, blockView)
+		}
+	}
+
+	return fmt.Errorf("block's view %d is not sequential %d, last view TC not included", blockView, qcView)
 }
 
 // IsSafeToTimeout checks if it's safe to timeout with proposed data, if timing out won't break safety rules.
-func (v *SafetyRules) IsSafeToTimeout(curView uint64, highestQC *flow.QuorumCertificate, highestTC *flow.TimeoutCertificate) bool {
-	panic("to be implemented")
+func (r *SafetyRules) IsSafeToTimeout(curView uint64, highestQC *flow.QuorumCertificate, highestTC *flow.TimeoutCertificate) bool {
+	if highestQC.View < r.safetyData.LockedOneChainView ||
+		curView+1 <= r.safetyData.HighestAcknowledgedView ||
+		curView <= highestQC.View {
+		return false
+	}
+
+	return (curView == highestQC.View+1) || (curView == highestTC.View+1)
```

Comment:
```suggestion
	if curView == highestQC.View+1 {
		return true
	}
	
	// sanity check: lastViewTC cannot be nil if last view did not result in QC
	if lastViewTC == nil { 
		return false
	}
	if curView == lastViewTC.View+1 {
		return true
	}
	return false
```

Suggestion to handle `lastViewTC` being unexpectedly nil.